### ISSUE TITLE
Unexpected dependencies?

### ISSUE THREAD

**Issue by toddr** [shawnlaffan/perl-alien-sqlite#2]
Hey, I noticed that you just released a fix for the tarball in the tarball. Thanks!

In the new release, I notice that these deps were added to META.json

      "Mojo::DOM" : "0",
      "Mojolicious" : "7.00",

Are they needed?

**Comment by shawnlaffan** [comment]
They're not explicitly listed in Makefile.PL.  I think they are pulled in by one of the Alien::Build plugins when I call ```perl Makefile.PL```.  

I'll see if I can trace which.

**Comment by shawnlaffan** [comment]
One source is Alien::Build::Plugin::Download::Negotiate

https://metacpan.org/source/PLICEASE/Alien-Build-2.21/lib/Alien/Build/Plugin/Download/Negotiate.pm#L50 

It can use several methods, so Mojo::DOM and Mojolicious are in some ways a side effect of what I have on my system.  

If you have a means of not listing them then I'd be happy to give it a try (preferably without manual intervention in the packaging process...).

**Comment by toddr** [comment]
I just wanted you to be aware. If it’s appropriate that’s fine. Thanks for your time

**Comment by toddr** [comment]
Ideally you should only be listing your direct dependencies. It’s up to the installers to figure out the chain

**Comment by shawnlaffan** [comment]
I agree, and none of Mojo::DOM58, Mojolicious or Mojo::DOM are explicitly listed in ```Makefile.PM```.  

The process seems to be that when I run ```perl Makefile.PL``` on my system it loads Alien::Build::Plugin::Download::Negotiate and adds additional libs to the makefile.  When I run ```gmake dist``` these are then listed in the meta files under build_requires.  They are also under ```x_alienfile.requires.share```, but not  ```x_alienfile.requires.system```.  

I don't know the internals of how makefiles are generated etc, but perhaps @plicease can shed some light on the process wrt Aliens.

**Comment by plicease** [comment]
TL;DR this is normal unpleasantness of authoring with EUMM (or MB for that matter).  

`Alien::Build` is aggressively dynamic in its prereqs so that deps aren't installed until they are determined necessarily for the local system.  `Mojo::DOM58` and `Mojolicious` are used to decode HTML for Aliens that need to decode an HTML file listing.  `Mojolicious` is preferred if it is already installed (why Install `Mojo::DOM58`, if `Mojolicious` will do the job?), otherwise we prefer `Mojo::DOM58` (because it is slightly lighter, and works on 5.8.x and 5.10.0 which AB must support).  We used to use `HTML::Parser` for this, but switched to the Mojo* options because they do not require XS or a C compiler.  When you install an Alien and AB determines that it needs one of these dynamic prereqs it inserts it int other prereqs for you.  This means we can improve the various AB plugins, without having to re-release dozens of Aliens.

Unfortunately, there isn't a way for AB to tell EUMM "I need this dependency for the local system, but not others."  So EUMM puts the dependencies that are needed on the system that builds the Alien dist.  When you run `Makefile.PL` on another system that doesn't need these deps, they won't be included in `MYMETA.json`, which is what cpan clients like `CPAN.pm` and `cpanm` actually use.  In short it all just works, but it does make static dependency analysis on metacpan harder.

If the Alien is authored using `Dist::Zilla` and the `[AlienBuild]` plugin, the static dependencies only are put in the `META.json`.  This works because the authoring tool and the installer are separate.  This is how I author Aliens, but it is probably not worth switching to dzil just for that.

**Comment by shawnlaffan** [comment]
Thanks for clarifying @plicease.  I guess this can be classified as "mostly harmless".

For comparison, Alien::proj does not list ```Mojo*``` under build_requires in its meta files, and the tarballs are built on the same machine I use for ```Alien::sqlite```.  However, ```Alien::proj``` is built as a system install by default on that machine.  Forcing a share install adds the relevant ```Mojo*``` libs to the meta files.  That would be consistent with your explanation.  

One nitpick with the explanation is wrt plugin order.  The code at https://metacpan.org/source/PLICEASE/Alien-Build-2.21/lib/Alien/Build/Plugin/Download/Negotiate.pm#L50 checks for ```Mojo::DOM58``` first, albeit both it and the ```Mojolicious``` check select ```Decode::Mojo```.  The ```Decode::Mojo``` plugin then checks first for ```Mojo::DOM58``` at https://metacpan.org/source/PLICEASE/Alien-Build-2.21/lib/Alien/Build/Plugin/Decode/Mojo.pm#L27

**Comment by shawnlaffan** [comment]
There's nothing left to be done with this issue, so I'll close it.  Thanks both for contributing.
